### PR TITLE
fix for issue #8636

### LINKED ISSUE #8636: Matches on &mut[] move the .. match & don't consider disjointness
I believe the following should work, since `head` and `tail` are disjoint, and `tail` is just taking a slice into `v` (this would mean that `v` is inaccessible while the result of `mut_head_tail` is kept around):

``` rust
fn mut_head_tail<'a, A>(v: &'a mut [A]) -> Option<(&'a mut A, &'a mut [A])> {
    match v {
        [ref mut head, .. tail] => {
            Some((head, tail))
        }
        [] => None
    }

    /* // this works currently:
    if v.is_empty() {
        None
    } else {
        let (low, hi) = v.mut_split(1);
        Some((&mut low[0], hi))
    }
    */
}

fn main() {
    let mut v = ~[1,2,3,4];

    match mut_head_tail(v) {
        None => {},
        Some((h,t)) => {
            *h = 1000;
            t.reverse();
        }
    }

    println!("{:?}", v);
}
```

```
mut-vec-match.rs:3:26: 3:30 error: cannot bind by-move and by-ref in the same pattern
mut-vec-match.rs:3         [ref mut head, .. tail] => {
                                             ^~~~
mut-vec-match.rs:3:9: 3:21 note: by-ref binding occurs here
mut-vec-match.rs:3         [ref mut head, .. tail] => {
                            ^~~~~~~~~~~~
error: aborting due to previous error
```

Changing it to (and removing the `t.reverse()`):

``` rust
fn mut_head_tail<'a, A>(v: &'a mut [A]) -> Option<(&'a mut A, ())> {
    match v {
        [ref mut head, .. ref mut _tail] => {
            Some((head, ()))
        }
        [] => None
    }
}
```

```
mut-vec-match.rs:21:9: 21:21 error: cannot borrow `(*v)[]` as mutable more than once at a time
mut-vec-match.rs:21         [ref mut head, .. ref mut _tail] => {
                             ^~~~~~~~~~~~
mut-vec-match.rs:21:8: 21:40 note: second borrow of `(*v)[]` as mutable occurs here
mut-vec-match.rs:21         [ref mut head, .. ref mut _tail] => {
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
```

### LINKED ISSUE #42291: Slice patterns don't work well with ref mut
Minimal example: https://is.gd/40m1am

```rust
#![feature(slice_patterns)]
fn main() {
    let mut buf = [0, 1];
    let [ref mut a, ref mut b] = buf;
}
```

This will complain about multiple mutable borrows, even though it should be valid.

### PR DESCRIPTION
**Pull request by mikhail-m1**
r? @nikomatsakis 

Fix #8636 

also fixes #42291

### CODE REVIEW

**Comment by mikhail-m1** [comment]
considering 
```
if let [_, ref subslice...] = slice {
    if let [ref a, ref b, ref c, ...] = subslice { ... }
}
```
I don't think additional tests are needed because `subslice` is still alive. Also `[a,[b,c]..]` syntax gives same MIR as `[a,b,c, ..].

File: src/librustc_mir/borrow_check/path_utils.rs [hunk 1]
```diff
@@ -381,30 +382,72 @@ fn place_element_conflict<'a, 'gcx: 'tcx, 'tcx>(
...
-                    debug!("place_element_conflict: DISJOINT-OR-EQ-ARRAY");
+                    debug!("place_element_conflict: DISJOINT-OR-EQ-ARRAY-INDEX");
+                    Overlap::EqualOrDisjoint
+                }
+                (ProjectionElem::ConstantIndex { offset: o1, min_length: _, from_end: false },
+                    ProjectionElem::ConstantIndex { offset: o2, min_length: _, from_end: false })
+                | (ProjectionElem::ConstantIndex { offset: o1, min_length: _, from_end: true },
+                    ProjectionElem::ConstantIndex { offset: o2, min_length: _, from_end: true }) =>
+                {
+                    if o1 == o2 {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/path_utils.rs]: Seems right =)

File: src/librustc_mir/borrow_check/path_utils.rs [hunk 2]
```diff
@@ -381,30 +382,72 @@ fn place_element_conflict<'a, 'gcx: 'tcx, 'tcx>(
...
+                        debug!("place_element_conflict: DISJOINT-ARRAY-CONSTANT-INDEX");
+                        Overlap::Disjoint
+                    }
+                }
+                (ProjectionElem::ConstantIndex { offset: offset_from_begin, min_length: min_length1, from_end: false },
+                    ProjectionElem::ConstantIndex { offset: offset_from_end, min_length: min_length2, from_end: true })
+                | (ProjectionElem::ConstantIndex { offset: offset_from_end, min_length: min_length1, from_end: true },
+                   ProjectionElem::ConstantIndex { offset: offset_from_begin, min_length: min_length2, from_end: false }) =>
+                {
+                    let length = max(min_length1, min_length2);
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/path_utils.rs]: Comment (this is my attempt):

---

Logic: we are matching a slice with *some* length dynamically; if both of these patterns matched, it must be at least the greater of the two.

File: src/librustc_mir/borrow_check/path_utils.rs [hunk 3]
```diff
@@ -381,30 +382,72 @@ fn place_element_conflict<'a, 'gcx: 'tcx, 'tcx>(
...
+                    if length <= offset_from_end || *offset_from_begin >= length - offset_from_end {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/path_utils.rs]: OK I must be confused about something here. =) Or else this logic is wrong. My money is on my being wrong, not the code. But in any case what would be nice I think are some comments and examples showing some interesting cases. Here is my attempt at an example, but it seems to suggest the code is wrong, so probably I am misunderstanding something:

---

- `[a, ..]` vs `[.., b]` -- here, for a list of length 1, `a` and `b` could both be the same element
  - `length` would be 1 (as min length in both cases is 1)
  - `offset_from_begin` and `offset_from_end` would be 0
  - but then we would return Disjoint?
    - `length <= offset_from_end` -- `1 <= 0`, false
    - `offset_from_begin >= length - offset_from_end` -- `0 >= 1 - 0`, false

---

What am I missing here?

File: src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+#![feature(slice_patterns)]
+
+fn nop(_s: &[& i32]) {}
+fn nop_subslice(_s: &[i32]) {}
+
+fn const_index_ok(s: &mut [i32]) {
+    if let [ref first, ref second, _, ref fourth, ..] = *s {
+        if let [_, _, ref mut third, ..] = *s {
+            nop(&[first, second, third, fourth]);
```
Comment (nikomatsakis) [inline src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs]: 👍

File: src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        }
+    }
+}
+
+fn const_index_err(s: &mut [i32]) {
+    if let [ref first, ref second, ..] = *s {
+        if let [_, ref mut  second2, ref mut third, ..] = *s { //~ ERROR cannot borrow `s[..]` as mutable because it is also borrowed as immutable
```
Comment (nikomatsakis) [inline src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs]: 👍

File: src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            nop(&[first, second, second2, third]);
+        }
+    }
+}
+
+fn const_index_from_end_ok(s: &mut [i32]) {
+    if let [.., ref fourth, ref third, _, ref first] = *s {
+        if let [.., ref mut second, _] = *s {
```
Comment (nikomatsakis) [inline src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs]: 👍

File: src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            nop(&[first, second, third, fourth]);
+        }
+    }
+}
+
+fn const_index_from_end_err(s: &mut [i32]) {
+    if let [.., ref fourth, ref third, _, ref first] = *s {
+        if let [.., ref mut third2, _, _] = *s { //~ ERROR cannot borrow `s[..]` as mutable because it is also borrowed as immutable
```
Comment (nikomatsakis) [inline src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs]: 👍

File: src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            nop(&[first, third, third2, fourth]);
+        }
+    }
+}
+
+fn const_index_and_subslice_ok(s: &mut [i32]) {
+    if let [ref first, ref second, ..] = *s {
+        if let [_, _, ref mut tail..] = *s {
```
Comment (nikomatsakis) [inline src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs]: 👍

File: src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            nop(&[first, second]);
+            nop_subslice(tail);
+        }
+    }
+}
+
+fn const_index_and_subslice_err(s: &mut [i32]) {
+    if let [ref first, ref second, ..] = *s {
+        if let [_, ref mut tail..] = *s { //~ ERROR cannot borrow `s[..]` as mutable because it is also borrowed as immutable
```
Comment (nikomatsakis) [inline src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs]: 👍

File: src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            nop(&[first, second]);
+            nop_subslice(tail);
+        }
+    }
+}
+
+fn const_index_and_subslice_from_end_ok(s: &mut [i32]) {
+    if let [.., ref second, ref first] = *s {
+        if let [ref mut tail.., _, _] = *s {
```
Comment (nikomatsakis) [inline src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs]: 👍

File: src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+            nop(&[first, second]);
+            nop_subslice(tail);
+        }
+    }
+}
+
+fn const_index_and_subslice_from_end_err(s: &mut [i32]) {
+    if let [.., ref second, ref first] = *s {
+        if let [ref mut tail.., _] = *s { //~ ERROR cannot borrow `s[..]` as mutable because it is also borrowed as immutable
```
Comment (nikomatsakis) [inline src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs]: 👍

File: src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+            nop(&[first, second]);
+            nop_subslice(tail);
+        }
+    }
+}
+
+fn subslices(s: &mut [i32]) {
+    if let [_, _, _, ref s1..] = *s {
+        if let [ref mut s2.., _, _, _] = *s { //~ ERROR cannot borrow `s[..]` as mutable because it is also borrowed as immutable
```
Comment (nikomatsakis) [inline src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs]: 👍

Comment (nikomatsakis) [reply src/test/compile-fail/borrowck/borrowck-slice-pattern-element-loan.rs]: (maybe we want a test like `[_, _, _, ref s1...]` vs `[ref mut s2.., _, _]` as well?)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks basically right and the tests are good. I'm having trouble following the logic but this mostly suggests to me we need some comments. =)

File: src/librustc_mir/borrow_check/places_conflict.rs [hunk 13]
```diff
@@ -394,27 +395,77 @@ fn place_element_conflict<'a, 'gcx: 'tcx, 'tcx>(
...
+                    offset: offset_from_begin, min_length: min_length1, from_end: false },
+                    ProjectionElem::ConstantIndex {
+                        offset: offset_from_end, min_length: min_length2, from_end: true })
+                | (ProjectionElem::ConstantIndex {
+                    offset: offset_from_end, min_length: min_length1, from_end: true },
+                   ProjectionElem::ConstantIndex {
+                       offset: offset_from_begin, min_length: min_length2, from_end: false }) => {
+                    // both patterns matched so it must be at least the greater of the two
+                    let min_length = max(min_length1, min_length2);
+                    // offset_from_end can be in range [1..min_length], -1 for last and min_length
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/places_conflict.rs]: Nit: it is not *-1* but rather just *1*...right?

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@bors r+

This is awesome! 💯

**Comment by nikomatsakis** [comment]
@bors r+

This is awesome! 💯

**Pull request merged**